New domain with human-readable name

orangejenny opened the pull request:
Change domain registration forms to accept a human-readable name, use that for domain.hr_name and convert it to a URL-friendly string for domain.name  A name like "My Project" will be translated to "my-project", or "my-project-2" if there are name conflicts.

Affects the new domain creation page and the exchange (when copying an exchange app into a new domain). UI looks the same, except for removing the "Project name cannot contain spaces" help text.

code buddy @proteusvacuum 

cc @amsagoff

orangejenny committed:
Use HR name when creating new domain, including from exchange

orangejenny committed:
Drop PDBs

orangejenny committed:
drop more debugging code

dimagimon commented on corehq/apps/appstore/views.py:
```diff
@@ -252,7 +256,7 @@ def copy_snapshot(request, domain):
 
             if form.is_valid():
-                new_domain = dom.save_copy(form.cleaned_data['domain_name'], user=user)
+                new_domain = dom.save_copy(form.cleaned_data['domain_name'], new_hr_name=form.cleaned_data['hr_name'], user=user)
```
E501 line too long (129 > 115 characters)

orangejenny committed:
dimagimon

czue commented on corehq/apps/registration/forms.py:
```diff
@@ -131,20 +131,27 @@ class DomainRegistrationForm(forms.Form):
+                counter = counter + 1
+                suffix = "-" + str(counter)
+                test_name = data[:self.max_name_length - len(suffix)] + suffix
+                conflict = Domain.get_by_name(test_name) or Domain.get_by_name(test_name.replace('-', '.'))
```
I think this should query the domain names once to find the next available key, not iterate like this hitting the DB every time until you get up to the right one.

orangejenny replied:
Changed to use ES. It's not down to one query for names with a lot of conflicts, but I figured that by the time we get to a point where there are that many conflicts, pulling in all domain names is going to be prohibitively expensive.

orangejenny committed:
don't crash on blank hr name

orangejenny committed:
Reduce queries done by name generation

dimagimon commented on corehq/apps/es/domains.py:
```diff
@@ -66,3 +67,7 @@ def is_active(is_active=True):
+
+def prefix(prefix):
+    return filters.prefix('name', prefix)
+
```
W391 blank line at end of file

orangejenny committed:
dimagimon

orangejenny committed:
add tests

dimagimon commented on corehq/apps/domain/tests/test_domain_name_generation.py:
```diff
@@ -0,0 +1,55 @@
+        for domain in self.domains:
+            domain.delete()
+
+class DomainNameGenerationBasicTest(DomainNameGenerationTestBase):
```
E302 expected 2 blank lines, found 1

orangejenny committed:
dimagimon

orangejenny committed:
wrong dimagimon

orangejenny committed:
Show validation error when name generation fails

orangejenny committed:
Stop calling generate_name twice

orangejenny committed:
update test

orangejenny committed:
comment out ES-dependent tests

czue commented:
hmm - I think ES actually comes with its own set of problems (not guaranteed to be 100% up to date).

haven't had a chance to look at this closely, but is there a reason you couldn't use the couch view with a start and endkey? that seems much safer.

benrudolph commented on corehq/apps/domain/tests/test_domain_name_generation.py:
```diff
@@ -0,0 +1,58 @@
+
+class DomainNameGenerationTestBase(TestCase):
+    def setUp(self):
+        self.client = Client()
```
took me a while to realize this, but django actually automatically instantiates `self.client` as `Client()`

orangejenny committed:
PR feedback

dimagimon commented on corehq/apps/domain/models.py:
```diff
@@ -21,6 +22,7 @@
 from corehq.apps.appstore.models import SnapshotMixin
 from corehq.util.quickcache import skippable_quickcache
 from corehq.util.dates import iso_string_to_datetime
+from corehq.apps.es.domains import DomainES
```
F401 'DomainES' imported but unused

dimagimon commented on corehq/apps/domain/tests/test_domain_name_generation.py:
```diff
@@ -0,0 +1,52 @@
+from __future__ import print_function, unicode_literals
+
+from django.test import TestCase
+from django.test.client import Client
```
F401 'Client' imported but unused

orangejenny committed:
drop unused imports

czue commented on corehq/apps/domain/models.py:
```diff
@@ -622,6 +623,47 @@ def get_or_create_with_name(cls, name, is_active=False,
+                name_prefix = name_prefix[:-shorten]
+            hits = Domain.get_names_by_prefix(name_prefix + "-")
+            if re.search(r'\.', name_prefix):
+                hits += Domain.get_names_by_prefix(name_prefix.replace('-', '.') + "-")
```
I think that this is here because we automatically map periods in domain names to hyphens? is that right?

seems worth a comment explaining the reasoning.

orangejenny replied:
I originally took that because that's how the orgs app checks for name conflicts: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/orgs/forms.py#L24

...but, re-reading, that's dealing with hr_name, not name, so I can drop this code.

czue commented on corehq/apps/domain/models.py:
```diff
@@ -622,6 +623,47 @@ def get_or_create_with_name(cls, name, is_active=False,
+                if match is not None:
+                    max_counter = max(max_counter, int(match.group(1)))
+            suffix = "-" + str(max_counter + 1)
+            name = name[:max_length - len(suffix)] + suffix
```
i trust that this logic works but it would be great to pull it out into a utility function and write some tests

czue replied:
e.g. if the function looks like this:
```python 
domain_name = find_next_available_name(my_name, list_of_taken_names)
```

you could have tests like this:
```python
def test_next_available():
    self.assertEqual('my-domain-3', find_next_available_name('my-domain', ['my-domain', 'my-domain-1', 'my-domain-2'])

def test_fits_in_a_gap():
    self.assertEqual('my-domain-3', find_next_available_name('my-domain', ['my-domain', 'my-domain-1', 'my-domain-2', 'my-domain-4'])
```

czue commented on corehq/apps/domain/tests/test_domain_name_generation.py:
```diff
@@ -0,0 +1,51 @@
+            domain.delete()
+
+
+class DomainNameGenerationBasicTest(DomainNameGenerationTestBase):
```
nice - I think you can split out the tests that are specifically exercising the different edge cases and 

1) for anything that doesn't have to depend on the DB pull out those tests and use `SimpleTestCase` so they run faster
2) for anything that does depend on the DB you can leave it (or if you're feeling especially motivated you can use mocks)

czue commented on corehq/apps/domain/models.py:
```diff
@@ -622,6 +623,47 @@ def get_or_create_with_name(cls, name, is_active=False,
+            name = name[:max_length - len(suffix)] + suffix
+
+        if has_conflict(name):
+            return None
```
any reason not to raise a `NoPossibleNameError` or similar? this seems like an exception-worthy failure mode

czue commented:
@orangejenny sorry about the misdirection on the past comments from partial review. this is looking good now and just some small feedback. also needs to be merged with master.

czue commented on corehq/apps/domain/models.py:
```diff
@@ -622,6 +623,47 @@ def get_or_create_with_name(cls, name, is_active=False,
+
+        name = original_name
+        shorten = -1
+        while has_conflict(name) and shorten + 2 < max_length:
```
possible pseudo-code for this:
```python
possible_conflicts = get_possible_conflicts(prefix, suffix_digits)
chosen_name = find_next_available_name(my_name, list_of_taken_names)
if chosen_name:
    break
else:
    suffix_digits, prefix = increment_how_far_back_i_look()
```

orangejenny committed:
drop ./- code

orangejenny committed:
Rearrange code based on PR feedback

dimagimon commented:
There are 10 errors:

* corehq/apps/domain/models.py, line 654 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 13 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 19 - E703 statement ends with a semicolon
* corehq/apps/domain/tests/test_domain_name_generation.py, line 23 - E703 statement ends with a semicolon
* corehq/apps/domain/tests/test_domain_name_generation.py, line 32 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 38 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 43 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 47 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 53 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 60 - E303 too many blank lines (2)

orangejenny committed:
merge master

dimagimon commented:
There are 11 errors:

* corehq/apps/appstore/views.py, line 265 - W291 trailing whitespace
* corehq/apps/domain/models.py, line 654 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 13 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 19 - E703 statement ends with a semicolon
* corehq/apps/domain/tests/test_domain_name_generation.py, line 23 - E703 statement ends with a semicolon
* corehq/apps/domain/tests/test_domain_name_generation.py, line 32 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 38 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 43 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 47 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 53 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 60 - E303 too many blank lines (2)

orangejenny committed:
docstring

dimagimon commented:
There are 11 errors:

* corehq/apps/appstore/views.py, line 265 - W291 trailing whitespace
* corehq/apps/domain/models.py, line 654 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 13 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 19 - E703 statement ends with a semicolon
* corehq/apps/domain/tests/test_domain_name_generation.py, line 23 - E703 statement ends with a semicolon
* corehq/apps/domain/tests/test_domain_name_generation.py, line 32 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 38 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 43 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 47 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 53 - E303 too many blank lines (2)
* corehq/apps/domain/tests/test_domain_name_generation.py, line 60 - E303 too many blank lines (2)

orangejenny commented:
@czue 
Split tests & rearranged the name-finding logic to be a bit clearer, lmk if it could use more improvement.

orangejenny committed:
dimagimon

orangejenny commented on corehq/apps/domain/models.py:
```diff
@@ -622,6 +623,48 @@ def get_or_create_with_name(cls, name, is_active=False,
+                name = Domain._get_next_available_name(prefix, Domain.get_names_by_prefix(prefix + '-'))
+                if Domain.get_by_name(name):
+                    # should never happen
+                    raise Exception
```
Open to using creating a custom exception if that's better practice. Given that it's only throw here and is immediately caught in forms.py, I took the lazier route.

czue replied:
I think good to add a custom exception. It's super easy to add and the benefit is that you don't end up swallowing other issues (like the database was down) and telling the user to keep trying new names.

orangejenny replied:
good point, fixed

orangejenny committed:
custom exception

dimagimon commented on corehq/apps/domain/exceptions.py:
```diff
@@ -1,2 +1,5 @@
 class InactiveTransferDomainException(Exception):
     pass
+
+class NameUnavailableException(Exception):
```
E302 expected 2 blank lines, found 1

dimagimon commented on corehq/apps/domain/tests/test_domain_name_generation.py:
```diff
@@ -0,0 +1,65 @@
+
+
+
+class DomainNameAvailabilityTest(SimpleTestCase):
```
E303 too many blank lines (3)

orangejenny committed:
dimagimon

czue commented on corehq/apps/domain/tests/test_domain_name_generation.py:
```diff
@@ -0,0 +1,64 @@
+
+        # Don't bother filling in gaps
+        similar.append("abc-9")
+        self.assertEquals(Domain._get_next_available_name(name, similar), "abc-10")
```
was this an intentional design decision or did it just seem like too much effort to support filling in gaps?

orangejenny replied:
Seemed like an edge case without negative consequences. Is there a reason you're concerned about it?

czue replied:
I think the "one user enters their year of birth and everyone else gets a huge number" one is the main thing, but I won't push too hard if you don't think it's worth it

czue commented on corehq/apps/domain/models.py:
```diff
@@ -622,6 +623,48 @@ def get_or_create_with_name(cls, name, is_active=False,
+            match = re.search(r'-([0-9]+)$', name)
+            if match:
+                max_suffix = max(max_suffix, int(match.group(1)))
+        return prefix + '-' + str(max_suffix + 1)
```
I think if you wanted to fill in gaps you can could just do something like:
```python

def _get_suffix_int(name):
  match = re.search(r'-([0-9]+)$', name)
  if match:
    return int(match.group(1))
  else:
    return None  # optional, for clarity

suffixes = sorted([_get_suffix_int(name) for name in existing_names])
for val, i in enumerate(suffixes):
  # i will be 0, 1, 2, 3.... the first time they aren't equal should be the first available gap
  if val != i:
    return i
```
(note: the above is not guaranteed to be correct)

orangejenny replied:
Ha, coincidentally, the gap-filling is one variation of one of our second-round interview questions, so there's a also a linear implementation of it described on confluence.

czue replied:
(i realized the above might break with negative numbers - not sure if the regex is equipped to deal with that or not)

czue commented on corehq/apps/domain/models.py:
```diff
@@ -662,8 +715,8 @@ def save(self, **params):
-    def save_copy(self, new_domain_name=None, user=None, ignore=None,
-                  copy_by_id=None):
+    def save_copy(self, new_domain_name=None, new_hr_name=None, user=None,
+                  ignore=None, copy_by_id=None):
```
(assume you looked for usages for this while changing argument order)

orangejenny replied:
yes

czue commented on corehq/apps/domain/tests/test_domain_name_generation.py:
```diff
@@ -0,0 +1,64 @@
+            domain.delete()
+
+    def test_generation(self):
+        self.assertEquals(Domain.generate_name("I have  spaces"), "i-have-spaces")
```
(not blocking) I think you could also pull out some of the space/lowercase logic into its own function and then make more of these tests not depend on the DB.

czue commented on corehq/apps/domain/models.py:
```diff
@@ -622,6 +623,48 @@ def get_or_create_with_name(cls, name, is_active=False,
+        return name
+
+    @classmethod
+    def _get_next_available_name(cls, prefix, existing_names):
```
(not blocking) I think this could just be a function rather than a `@classmethod`. if moving to a function I would also suggest making it a public helper (so removing the leading `_`). I think we generally try to make `_` mean private even in tests, though that's not a hard requirement but more of a guideline. I think at least pycharm will complain when you access this in tests though.

orangejenny replied:
Can you explain more about why this should be a function? It doesn't depend on any internal state, and it isn't really a reusable piece of functionality. Not that it's impossible anyone would ever reuse it, but I prefer to keep functionality private until there's a reason for it to be public. As is, it allows for tests without hitting the db, but developers know it isn't recommended for use outside this class.

orangejenny replied:
Looking around some more, would it more typical to put `generate_name` and `_get_next_available_name` into utils.py? (although then I don't imagine I'd be able to mark `_get_next_available_name` private with a straight face)

orangejenny replied:
Although, the notion of a private class method doesn't make sense, so I should probably change this somehow. Making a logical API for Domain, with public/private determined by how other devs should use the code, and writing code that can be tested granularly seem a bit at odds here.

czue replied:
yeah I was advocating for public somewhere (utils sounds fine)

I think most of our difference of opinion is generally coming from a slightly different mindset about these:

- you're thinking the're super-specific helper functions for this case
- i'm thinking of the general "choose an auto-generated readable name based on some size constraints and incrementing something"  function

I don't feel strongly enough to think I'm right that it's general, though I don't know that I see harm in thinking of it that way.

czue commented:
my preference would be to try harder to fill in gaps here (if I make a domain called test-9999 does that mean that everyone who tries "test" first after me gets test-1000+?). rest are non-blockers.

a few other questions i had:

- Have we tested that the form where you put this in blocks double submission? I think there's still an edge case/race condition where you generate the name and then by the time you go and save it there's a duplicate - though I think that also exists today.
- Does this need any QA or UAT or is it good to go when code review is done?

snopoke commented:
Could use a critical section to prevent the double submission issue.

On 23 June 2015 at 00:08, Cory Zue <notifications@github.com> wrote:

> my preference would be to try harder to fill in gaps here (if I make a
> domain called test-9999 does that mean that everyone who tries "test" first
> after me gets test-1000+?). rest are non-blockers.
>
> a few other questions i had:
>
>    - Have we tested that the form where you put this in blocks double
>    submission? I think there's still an edge case/race condition where you
>    generate the name and then by the time you go and save it there's a
>    duplicate - though I think that also exists today.
>    - Does this need any QA or UAT or is it good to go when code review is
>    done?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/7106#issuecomment-114281845>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

orangejenny commented:
@czue 

> Does this need any QA or UAT or is it good to go when code review is done?

I don't think it needs UAT. I've manually tested it, and obviously it has tests, but yeah, having more than one person run through new functionality is a good thing. Will make a QA ticket once this is ready to be pushed to staging.

orangejenny commented:
@czue That is, I don't think UAT feedback is likely to be useful, because the user interaction isn't  changing.

orangejenny committed:
separate out normalization logic for db-independent testing

czue commented:
I think i'm signed off on this with a few "I think it should be different but don't feel strongly enough to block merge" comments

orangejenny committed:
Use CriticalSection

dimagimon commented:
There are 100 errors:

* corehq/apps/domain/models.py, line 733 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 733 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 735 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 735 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 737 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 737 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 738 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 738 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 739 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 739 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 740 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 740 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 741 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 741 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 742 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 742 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 743 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 743 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 744 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 744 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 746 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 746 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 747 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 747 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 748 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 748 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 749 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 749 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 750 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 750 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 751 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 751 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 752 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 752 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 753 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 753 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 754 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 754 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 756 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 756 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 757 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 757 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 758 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 758 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 760 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 760 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 762 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 762 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 763 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 763 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 764 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 764 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 765 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 765 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 766 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 766 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 767 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 767 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 769 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 769 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 770 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 770 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 771 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 771 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 772 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 772 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 773 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 773 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 774 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 774 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 775 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 775 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 776 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 776 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 777 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 777 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 778 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 778 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 779 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 779 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 779 - E501 line too long (119 > 115 characters)
* corehq/apps/domain/models.py, line 780 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 780 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 781 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 781 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 782 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 782 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 783 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 783 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 784 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 784 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 785 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 785 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 787 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 787 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 788 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 788 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 790 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 790 - E101 indentation contains mixed spaces and tabs
* corehq/apps/registration/forms.py, line 7 - F401 'NameUnavailableException' imported but unused

orangejenny committed:
move functions to utils

orangejenny committed:
...and include utils file

dimagimon commented:
There are 101 errors:

* corehq/apps/domain/models.py, line 5 - F401 're' imported but unused
* corehq/apps/domain/models.py, line 717 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 717 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 719 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 719 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 721 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 721 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 722 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 722 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 723 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 723 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 724 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 724 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 725 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 725 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 726 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 726 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 727 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 727 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 728 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 728 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 730 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 730 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 731 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 731 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 732 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 732 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 733 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 733 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 734 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 734 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 735 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 735 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 736 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 736 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 737 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 737 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 738 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 738 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 740 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 740 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 741 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 741 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 742 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 742 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 744 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 744 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 746 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 746 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 747 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 747 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 748 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 748 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 749 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 749 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 750 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 750 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 751 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 751 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 753 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 753 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 754 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 754 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 755 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 755 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 756 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 756 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 757 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 757 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 758 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 758 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 759 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 759 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 760 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 760 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 761 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 761 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 762 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 762 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 763 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 763 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 763 - E501 line too long (119 > 115 characters)
* corehq/apps/domain/models.py, line 764 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 764 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 765 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 765 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 766 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 766 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 767 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 767 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 768 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 768 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 769 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 769 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 771 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 771 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 772 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 772 - E101 indentation contains mixed spaces and tabs
* corehq/apps/domain/models.py, line 774 - W191 indentation contains tabs
* corehq/apps/domain/models.py, line 774 - E101 indentation contains mixed spaces and tabs
* corehq/apps/registration/forms.py, line 7 - F401 'NameUnavailableException' imported but unused

orangejenny committed:
Merge branch 'master' into new-domain-with-hr-name

orangejenny committed:
vim is not my friend

dimagimon commented on corehq/apps/domain/models.py:
```diff
@@ -2,6 +2,7 @@
 from itertools import imap
 import json
 import logging
+import re
```
F401 're' imported but unused

dimagimon commented on corehq/apps/domain/models.py:
```diff
@@ -662,74 +702,76 @@ def save(self, **params):
+                    if app:
+                        comp = self.copy_component(app.doc_type, app._id, new_domain_name, user=user)
+                    else:
+                        comp = self.copy_component(res['value']['doc_type'], res['value']['_id'], new_domain_name, user=user)
```
E501 line too long (125 > 115 characters)

dimagimon commented on corehq/apps/registration/forms.py:
```diff
@@ -4,10 +4,11 @@
 from django.contrib.auth.models import User
 from corehq.apps.users.forms import RoleForm, SupplyPointSelectWidget
 import re
+from corehq.apps.domain.exceptions import NameUnavailableException
```
F401 'NameUnavailableException' imported but unused

orangejenny committed:
move functions from domain.utils to dimagi-utils

dimagimon commented on corehq/apps/domain/tests/test_domain_name_generation.py:
```diff
@@ -0,0 +1,45 @@
+from __future__ import print_function, unicode_literals
+
+from django.test import SimpleTestCase, TestCase
```
F401 'SimpleTestCase' imported but unused

dimagimon commented on corehq/apps/domain/tests/test_domain_name_generation.py:
```diff
@@ -0,0 +1,45 @@
+
+from corehq.apps.domain.exceptions import NameUnavailableException
+from corehq.apps.domain.models import Domain
+from dimagi.utils.name_to_url import name_to_url
```
F401 'name_to_url' imported but unused

dimagimon commented on corehq/apps/domain/tests/test_domain_name_generation.py:
```diff
@@ -0,0 +1,45 @@
+from corehq.apps.domain.exceptions import NameUnavailableException
+from corehq.apps.domain.models import Domain
+from dimagi.utils.name_to_url import name_to_url
+from dimagi.utils.next_available_name import next_available_name
```
F401 'next_available_name' imported but unused

orangejenny committed:
drop unused imports

orangejenny committed:
one more unused import

orangejenny commented:
@czue / @snopoke 

Moved name-making helper functions into dimagi-utils, added CriticalSections around domain name generation + saving, left the name generation code not filling in gaps.

czue commented:
@orangejenny tests are failing (looks like it might be a circular import problem?)
```
Traceback (most recent call last):
  File "manage.py", line 65, in <module>
    execute_from_command_line(sys.argv)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 399, in execute_from_command_line
    utility.execute()
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 392, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 272, in fetch_command
    klass = load_command_class(app_name, subcommand)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 75, in load_command_class
    module = import_module('%s.management.commands.%s' % (app_name, name))
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/utils/importlib.py", line 40, in import_module
    __import__(name)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/management/commands/__init__.py", line 10, in <module>
    import django.template.loaders.app_directories
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/template/loaders/app_directories.py", line 25, in <module>
    raise ImproperlyConfigured('ImportError %s: %s' % (app, e.args[0]))
ImproperlyConfigured: ImportError corehq.apps.userhack: No module named name_to_url
```

czue commented:
oh, just kidding, it looks like a dimagi-utils submodule ref update needed

orangejenny committed:
Merge branch 'master' into new-domain-with-hr-name

orangejenny committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into new-domain-with-hr-name

dimagimon commented on corehq/apps/locations/views.py:
```diff
@@ -59,6 +59,10 @@ def dispatch(self, request, *args, **kwargs):
     def section_url(self):
         return reverse(LocationsListView.urlname, args=[self.domain])
 
+    @method_decorator(locations_access_required)
```
F811 redefinition of unused 'dispatch' from line 54

proteusvacuum committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into new-domain-with-hr-name

orangejenny committed:
fix tests

orangejenny commented:
@czue fixed tests, passed qa, should be good to go

czue commented:
:tada:
